daemonize.c

```diff
@@ -22,85 +22,167 @@
 		}											\
 		fprintf(stderr, "%s\n", strerror(errno));	\
 		exit(EXIT_FAILURE);							\
 	} while (0)										
 
-static void print_usage();
-
 static char *name;
+static char *version = "v0.1";
 
-int main(int argc, char **argv){
-	int fd;
+static void print_usage(char **, struct option const *);
+
+
+int main(int argc, char **argv) {
+	int nullfd, outredfd, errredfd;
 	int old_errno;
 	pid_t child;
 	struct sigaction sa;
 
+	int retopt = -1;
+	struct {
+		char command[NAME_MAX];
+		char outred[NAME_MAX];
+		char errred[NAME_MAX];
+
+		char *const *arguments;
+	} pargs = {"", DEVNULL, DEVNULL, NULL};
+
+	int optindex;
+	char const *shortopts = "+ h v o: e:";
+	const struct option longopts[] = {
+		{"out", required_argument, NULL, OPT_OUTRED},
+		{"err", required_argument, NULL, OPT_ERRRED},
+		{"help", no_argument, NULL, OPT_HELP},
+		{"version", no_argument, NULL, OPT_VERSION},
+	};
+
 	name = argv[0];
-	
-	if (argc < 2){
-		print_usage();
-		return EXIT_SUCCESS;
+
+	do {
+		retopt = getopt_long(argc, argv, shortopts, longopts, &optindex);
+
+		switch (retopt) {
+			case OPT_OUTRED:
+				strncpy(pargs.outred, optarg, NAME_MAX);
+				break;
+			case OPT_ERRRED:
+				strncpy(pargs.errred, optarg, NAME_MAX);
+				break;
+			case OPT_HELP:
+				print_usage(argv, longopts);
+				exit(EXIT_SUCCESS);
+			case OPT_VERSION:
+				puts(version);
+				exit(EXIT_SUCCESS);
+			case -1:
+			default:
+				break;
+		}
+	} while (retopt > -1);
+
+	/*
+	 * Consume the remaining non-option arguments from getopt_long().
+	 * TO-DO: here is what the man-page reads about argv:
+	 * «[...] The first argument, by convention, should point to the filename
+	 * associated with the file being executed.  The array of pointers must be
+	 * terminated by a null pointer. [...]»
+	 * So shall `pargs.arguments` actually point to the name of the file, and
+	 * not just the argument(s) coming next?
+	 */
+	// Copy the command filename from the first remaining non-option argument
+	if (optind < argc) {
+		strncpy(pargs.command, argv[optind], NAME_MAX);
 	}
-	
-	if ((child = fork()) == 0){
+	// If other non-option argument values are left besides the filename, we
+	// add them to pargs.arguments (whilst also including the filename value,
+	// hence pargs.command == pargs.arguments if the if-condition holds true)
+	if (optind + 1 < argc) {
+		pargs.arguments = argv + optind;
+	}
+
+	if ((child = fork()) == 0) {
 		/* Ignore SIGHUP */
 		memset(&sa, 0, sizeof(sa));
 		sa.sa_handler = SIG_IGN;
 		sigaction(SIGHUP, &sa, NULL);
 
 		/* Create a new session */
-		if (setsid() < 0){
+		if (setsid() < 0) {
 			errx("can't create a new session");
 		}
 
-		/* Redirect I/O to /dev/null open for writing */
-		if ((fd = open(DEVNULL, O_WRONLY)) < 0){
+		/* Redirect input to /dev/null open for writing */
+		if ((nullfd = open(DEVNULL, O_WRONLY)) < 0)
 			errx("can't open %s", DEVNULL);
-		}
-
-		if (isatty(STDIN_FILENO)){
-			dup2(fd, STDIN_FILENO);
-		}
+		if (isatty(STDIN_FILENO))
```

Comment:
With @forsenonlhaimaisentito hopefully approving, I've weeded out some of the braces here (there were too many).